### ISSUE TITLE
High CPU usage by php-fpm process

### ISSUE THREAD

**Issue by andybalaam** [nextcloud-snap/nextcloud-snap#1586]
#### Describe the bug
After installing this snap my CPU usage is high.  According to `top`, several `php-fpm` processes often use over 90% of the CPU, and the load average is often over 2.0.

My system has 2 AMD CPUs and 4GB RAM.  Full details are at the bottom.

#### To Reproduce
1. Uninstall the snap
2. Measure CPU load average with `top` or `uptime`
2. Install the snap
3. Measure CPU load average with `top` or `uptime` again
4. Note that load average is much higher

#### Expected behavior
I would expect to be able to configure the snap to use less CPU on my relatively modest system

#### OS/snapd/snap version

```bash
$ snap list nextcloud
Name       Version                   Rev    Tracking       Publisher   Notes
nextcloud  20.0.3snap1+git1.f36af40  24967  latest/beta/…  nextcloud✓  -
$ snap version
snap    2.48
snapd   2.48
series  16
ubuntu  20.04
kernel  5.4.0-56-generic
$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.1 LTS"
```

#### CPU Info:

```bash
$ cat /proc/cpuinfo
processor	: 0
vendor_id	: AuthenticAMD
cpu family	: 20
model		: 2
model name	: AMD G-T40E Processor
stepping	: 0
microcode	: 0x5000119
cpu MHz		: 999.892
cache size	: 512 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 2
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 6
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni monitor ssse3 cx16 popcnt lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch ibs skinit wdt hw_pstate vmmcall arat npt lbrv svm_lock nrip_save pausefilter
bugs		: fxsave_leak sysret_ss_attrs null_seg spectre_v1 spectre_v2 spec_store_bypass
bogomips	: 1999.76
TLB size	: 1024 4K pages
clflush size	: 64
cache_alignment	: 64
address sizes	: 36 bits physical, 48 bits virtual
power management: ts ttp tm stc 100mhzsteps hwpstate

processor	: 1
vendor_id	: AuthenticAMD
cpu family	: 20
model		: 2
model name	: AMD G-T40E Processor
stepping	: 0
microcode	: 0x5000119
cpu MHz		: 999.891
cache size	: 512 KB
physical id	: 0
siblings	: 2
core id		: 1
cpu cores	: 2
apicid		: 1
initial apicid	: 1
fpu		: yes
fpu_exception	: yes
cpuid level	: 6
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni monitor ssse3 cx16 popcnt lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch ibs skinit wdt hw_pstate vmmcall arat npt lbrv svm_lock nrip_save pausefilter
bugs		: fxsave_leak sysret_ss_attrs null_seg spectre_v1 spectre_v2 spec_store_bypass
bogomips	: 1999.76
TLB size	: 1024 4K pages
clflush size	: 64
cache_alignment	: 64
address sizes	: 36 bits physical, 48 bits virtual
power management: ts ttp tm stc 100mhzsteps hwpstate
```
Memory usage seens fine:

```bash
$ free
              total        used        free      shared  buff/cache   available
Mem:        3619632      895656      139700       28692     2584276     2409576
Swap:       3904508      133376     3771132
```

**Comment by kyrofa** [comment]
The initial bootstrapping process (when initially installing the snap) is relatively high-cost. Do you find the CPU keeps churning after Nextcloud is installed, when it should be idling?

**Comment by andybalaam** [comment]
Yes, this has been running for months.  I recently added a lot more files and clients, and that is when it became CPU-hungry.

**Comment by andybalaam** [comment]
I _think_ it could be triggered by clients polling for updates, but I'm not sure.

**Comment by kyrofa** [comment]
Yeah that would imply that it's being actively used. As far as the PHP configuration is concerned, the snap tries to scale to the hardware it's on. In that sense, using a significant portion of the CPU when it's being used sounds like it's working the way it's supposed to. If it used less CPU your clients would get a worse experience. If, however, you're wanting to use that server for something else as well, and you need the extra headroom, you can scale the snap back by throwing it in a VM or LXD container that presents a different system to the snap, causing it to scale differently.

Other than that, there are [interesting discussions happening around snapd growing some relevant features](https://forum.snapcraft.io/t/limiting-snaps-memory-usage/11919). We can talk about ways to make the snap more configurable in that regard, but I'm not looking to just throw config files in a writable location and say "have at it." We would need to design a nice UX for this.

**Comment by andybalaam** [comment]
I should have mentioned that the behaviour on clients was very slow, taking hours to check for remote changes, so I would not characterise this as simply "working hard to service clients" but rather "over-using resources resulting in poor performance".

**Comment by andybalaam** [comment]
In case others have similar problems, I will tentatively mention what I did that seemed to improve the situation.

THIS IS NOT RECOMMENDED BY ANYONE, AND I DON'T KNOW HOW BADLY THIS IMPACTS SECURITY!

That said, here is what I did:

```
sudo nextcloud.occ config:system:set hashingThreads --type=integer --value=1
sudo nextcloud.occ config:system:set hashingTimeCost --type=integer --value=1
sudo nextcloud.occ config:system:set hashingMemoryCost --type=integer --value=8
```

This reduces the complexity of the password hashing, which reduces CPU usage.

After making the above change, CPU usage seems lower, and the clients seem to sync faster.
